sourcerer-core/src/main/java/org/elder/sourcerer2/DefaultCommand.kt

```diff
@@ -0,0 +1,235 @@
+            aggregate = null
+        }
+
+        // Bail out early if idempotent create, and already present
+        if (idempotentCreate
+                && aggregate != null
+                && aggregate.sourceVersion() != Aggregate.VERSION_NOT_CREATED
+        ) {
+            logger.debug("Bailing out early as already created (and idempotent create set)")
+            return CommandResult(
+                    aggregateId,
+                    aggregate.sourceVersion(),
+                    ImmutableList.of())
+        }
+
+        // Execute the command handler
+        val operationEvents = operation.execute(aggregate, arguments)
+        val events = ImmutableList.copyOf(operationEvents.stream().iterator())
+
+        if (events.isEmpty()) {
+            logger.debug("Operation is no-op, bailing early")
+            return CommandResult(
+                    aggregateId,
+                    aggregate?.sourceVersion(),
+                    events)
+        }
+
+        // Create/update the event stream as needed
+        val updateExpectedVersion = if (atomic) {
+            // Actually null safe since atomic above ...
+            if (aggregate!!.sourceVersion() !== Aggregate.VERSION_NOT_CREATED) {
+                ExpectedVersion.exactly(aggregate!!.sourceVersion())
+            } else {
+                ExpectedVersion.notCreated()
+            }
+        } else if (idempotentCreate) {
+            ExpectedVersion.notCreated()
+        } else {
+            ExpectedVersion.any()
+        }
+
+        logger.debug("About to persist, expected version at save: {}", updateExpectedVersion)
+        val effectiveMetadata = HashMap(this.metadata)
+        for (metadataDecorator in metadataDecorators) {
+            val decoratorMetadata = metadataDecorator.metadata
+            if (decoratorMetadata != null) {
+                effectiveMetadata.putAll(decoratorMetadata)
+            }
+        }
+
+        effectiveMetadata.putAll(this.metadata)
+
+        try {
+            val newVersion = repository.append(
+                    aggregateId!!,
+                    events,
+                    updateExpectedVersion,
+                    effectiveMetadata
+            )
+
+            logger.debug("Save successful, new version is {}", newVersion)
+            return CommandResult(aggregateId, newVersion, events)
+        } catch (ex: UnexpectedVersionException) {
+            // There's one case when this is OK - idempotent creates. We want to be able to create
+            // a stream and not fail if the same stream is attempted to be created on replays.
+            if (idempotentCreate) {
+                logger.debug("Idempotent create enabled, ignoring existing stream")
+                return CommandResult(
+                        aggregateId,
+                        if (ex.currentVersion != null) ex.currentVersion else null,
+                        ImmutableList.of<TEvent>())
+            }
+
+            throw ex
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Do not rethrow a caught exception of the same type [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.elder.org/coding_rules#rule_key=detekt-kotlin%3ARethrowCaughtException)